### ISSUE TITLE
Option to preprocess input and variables like {area} before matching.

### ISSUE THREAD

**Issue by Arkkimaagi** [home-assistant/intents#907]
# Sorry, long post, here's the beef:

What if it would be possible to preprocess the user input to convert the words to their base form (lemmatize) for easier matching with complex languages?

## Background

So, I started to write sentences for Finnish and I hit a wall quite fast. Finnish is [hard](https://www.studyfinnish.com/images/polandball/Dog.png).

### Generating variations in _common.yaml

First I tried declinating the area strings to all their different versions. Sure, this produces plenty of false words and noise, but it gets many right too and it's available currently.

```yaml
  kerro: "(kerro|näytä|sano)" # tell / show / say
  paljonko: "(paljonko|mikä)" # how much / what
  lämpötila: "(lämpötila|lämpö|sisälämpötila|ulkolämpötila|asteet)" # temperature / temp / indoor temperature / outdoor temperature / degrees
  nyt: "(nyt|juuri nyt|just nyt|tällä hetkellä)" # now / right now / rightabout now / at this moment
  vokaali: "(a|e|i|o|u|y|ä|ö)" # Finnish vovels

  # Area declinations
  huone: "{area}" # nominatiivi = mikä, kuka
  huoneen: "{area}[<vokaali>](n|en|in|den|ten|tten)" # genetiivi = minkä, kenen
  huonetta: "{area}[<vokaali>](a|ä|ta|tä|tta)" # partitiivi = mitä, ketä
  huoneena: "{area}[<vokaali>](na|nä)" # essiivi = minä, kenä
  huoneeksi: "{area}[<vokaali>](ksi|kse)" # translatiivi = miksi, keneksi
  huoneessa: "{area}[<vokaali>](ssa|ssä)" # inessiivi = missä, kenessä
  huoneesta: "{area}[<vokaali>](sta|stä)" # elatiivi = mistä, kenestä
  huoneeseen: "{area}[<vokaali>]((a|e|i|o|u|y|ä|ö)n|h(a|e|i|o|u|y|ä|ö)n|seen|siin)" # illatiivi = mihin, keneen
  huoneella: "{area}[<vokaali>](lla|llä)" # adessiivi = millä, kenellä
  huoneelta: "{area}[<vokaali>](lta|ltä)" # ablatiivi = miltä, keneltä
  huoneelle: "{area}[<vokaali>]lle" # allatiivi = mille, kenelle
  huoneetta: "{area}[<vokaali>](tta|ttä)" # abessiivi = ilman mitä, ketä
  huonein: "{area}[<vokaali>]n" # instruktiivi = miten
  huoneineen: "{area}[<vokaali>]ine" # komitatiivi = minkä, kenen kanssa
 ```

I then tried to use these in sentences. I just put the "tell", "room" (with proper declinations) "temperature" "now" words in different order each row and checked what variants of area name were grammatically okay for that order.

```yaml
          - "[<kerro>] (<huone>|<huoneen>) <lämpötila>"
          - "[<kerro>] (<huone>|<huoneen>|<huoneesta>) <lämpötila> <nyt>"
          - "[<kerro>] (<huone>|<huoneesta>|<huoneelta>|<huoneelle>) <nyt> <lämpötila>"
          - "[<kerro>] <lämpötila> (<huone>|<huoneessa>|<huoneesta>|<huoneelle>) [<nyt>]"
          - "[<kerro>] <lämpötila> <nyt> (<huone>|<huoneessa>|<huoneesta>|<huoneelle>)"
          - "[<kerro>] (<nyt>|nykyinen) (<huone>|<huoneen>) <lämpötila>"
          - "[<kerro>] (<nyt>|nykyinen) <lämpötila> (<huone>|<huoneessa>|<huoneesta>|<huoneelle>)"
```

I ended up with 134 rows of different sentences like this and more would be coming. In Finnish one can arrange the words in pretty much all the combinations and only a few variations change the meaning drastically.

The problem with this approach is that it's sloooow. And I tested how many unique variations it generates, I got around 4.5 million rows with my 12 areas and 134 sentences to ask for the temperature. That file took 450 megs of space.

### Generating variations by users as area aliases

Abandoning this approach I changed all my `(<huone>|<huoneesta>|<huoneelta>|<huoneelle>)` bits from the sentences back to single `{area}`

Then I added only the declinations currently in use in sentences to fixtures like this:

```yaml
language: "fi"
areas:
  - name: "(keittiö|keittiön|keittiössä|keittiöstä|keittiöön|keittiöllä|keittiöltä|keittiölle)"
    id: "kitchen"
  - name: "(olohuone|olohuoneen|olohuoneessa|olohuoneesta|olohuoneeseen|olohuoneella|olohuoneelta|olohuoneelle)"
    id: "living_room"
  - name: "(makuuhuone|makuuhuoneen|makuuhuoneessa|makuuhuoneesta|makuuhuoneeseen|makuuhuoneella|makuuhuoneelta|makuuhuoneelle)"
    id: "bedroom"
```

This reduces the spam significantly, but it's still slow. It's still hundreds of thousands of variations just for asking the freaking temperature.

## Suggestion

What if instead of generating all these different variations for finnish we'd reduce all input and sentence words to their base format? This should reduce the amount of comparisons drastically. So, all the sentences would be run through this process, and the user input would be too. Then these simple, base formats would be compared for a match.

I wrote this script:

https://gist.github.com/Arkkimaagi/5b5f40ecb846910d8ffd71bcfdbeddff

It converts Finnish words to their base form.

If I run the sentences example from first approach through it, it returns this:
```yaml
          - "[<kertoa>] (<huone>|<huone>) <lämpötila>"
          - "[<kertoa>] (<huone>|<huone>|<huone>) <lämpötila> <nyt>"
          - "[<kertoa>] (<huone>|<huone>|<huone>|<huone>) <nyt> <lämpötila>"
          - "[<kertoa>] <lämpötila> (<huone>|<huone>|<huone>|<huone>) [<nyt>]"
          - "[<kertoa>] <lämpötila> <nyt> (<huone>|<huone>|<huone>|<huone>)"
          - "[<kertoa>] (<nyt>|nykyinen) (<huone>|<huone>) <lämpötila>"
          - "[<kertoa>] (<nyt>|nykyinen) <lämpötila> (<huone>|<huone>|<huone>|<huone>)"
```

Which can be simplified to this:

```yaml
          - "[<kertoa>] {area} <lämpötila>"
          - "[<kertoa>] {area} <lämpötila> <nyt>"
          - "[<kertoa>] {area} <nyt> <lämpötila>"
          - "[<kertoa>] <lämpötila> {area} [<nyt>]"
          - "[<kertoa>] <lämpötila> <nyt> {area}"
          - "[<kertoa>] (<nyt>|nykyinen) {area} <lämpötila>"
          - "[<kertoa>] (<nyt>|nykyinen) <lämpötila> {area}"
```

And with user input like this: "Kerro lämpötila olohuoneessa nyt"
My script returns this: "kertoa lämpötila olohuone nyt"

My name for area knwon as living_room is "olohuone", that could be run through the same code too, but by default I'd imagine the area names are in their base form on most users setups like mine.

So by stripping the words to their base form, we'd be able to match them with a way smaller cpu and memory load. I have not been able to test it yet, but I'd imagine that this would reduce the load from Finnish conjugations and declinations a ton.

### Bonus:

If we'd want to go even further, lets have a syntax for saying "If these (base) words are in any order at this spot, consider it a match". 

Like if I write 

```yaml
          - "^word1|word2|[word3]^"
```
It would mean the same as:
```yaml
          - "word1 word2 [word3]"
          - "word1 [word3] word2"
          - "word2 word1 [word3]"
          - "word2 [word3] word1"
          - "[word3] word1 word2"
          - "[word3] word2 word1"
```
That could reduce the sentences from above suggestion to something like:

```yaml
          - "[<kertoa>] ^{area}|<lämpötila>|[<nyt>]^"
```

## TLDR

Could we please have an option to run code for the user input and variable strings (area, device_name, etc) before matching?

**Comment by tetele** [comment]
As discussed on Discord, this seems like the perfect use case for https://github.com/home-assistant/hassil/pull/30

**Comment by salleq** [comment]
I don't know if this helps even a bit with the different (declinations? not sure about the term), but for now I approached the issue by having different aliases in HA for different endings of the name of the room. For instance, I have the following aliases for olohuone (living room):

- olohuone
- olohuoneessa
- olohuoneesta
- olohuoneen

So in this case I could say the following sentences (rought translation to English in  parentheses to see differences):

1. sammuta valot huoneessa olohuone (turn off lights in room "olohuone")
2. sammuta valot olohuoneessa (turn off lights in "olohuone")
3. sammuta valot olohuoneesta (turn off lights from "olohuone")
4. sammuta olohuoneen valot (turn off lights of "olohuone")

Obviously this is something that each user has to do for every room once and as such isn't optimal, but not all declinations(?) are needed for each room, for instance "sammuta valot olohuoneella/olohuoneelta (turn off light at "olohuone")" is something that maybe some Finnish dialects might use, but are not generally used.

But like I said, I don't know if approaching the issue from alias side helps or not...
